OSDOCS-19410-5 back porting most recent zero-trust docs to 4.18 PR#5 (#110970)
openshift/openshift-docs: modules/zero-trust-manager-federation-configuration.adoc

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-spire-federation.adoc

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-federation-configuration_{context}"]
= Configuring SPIRE federation with the https_spiffe profile

[role="_abstract"]
The {zero-trust-full} includes SPIRE Federation support, allowing multiple independent SPIRE deployments to establish trust relationships. This procedure demonstrates how to configure federation using the `https_spiffe` profile, which uses SPIFFE-based TLS authentication between SPIRE servers.

.Prerequisites

> * You installed the OpenShift CLI (`oc`).

Reviewer: ```suggestion
* You installed the {oc-first}.
```
and no spaces between bullets

Author: Done

* You installed the {zero-trust-full} on all clusters that participate in the federation.

* You have `cluster-admin` privileges on all participating clusters.

* You have network connectivity between the clusters you intend to federate.

.Procedure

. Configure the `SpireServer` custom resource on each cluster to enable federation with the `https_spiffe` profile. The `https_spiffe` profile uses SPIFFE-based TLS authentication, where SPIRE servers authenticate to each other using their own SPIFFE Verifiable Identity Documents (SVIDs).
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: SpireServer
metadata:
  name: cluster
spec:
  trustDomain: cluster1.example.com
  federation:
    bundleEndpoint:
      profile: https_spiffe
      refreshHint: 300
    managedRoute: "true"
----

> 

Reviewer: ```suggestion
+
```

Author: Done

* The `spec.trustDomain` field sets a unique trust domain for each cluster.

> 

Reviewer: ```suggestion
```
etc

Author: Done

* The `spec.federation.bundleEnpoint.profile` field uses the `https_spiffe` profile for SPIFFE-based TLS authentication.

* The `spec.federation.bundleEndpoint.refreshHint` field suggests intervals (in seconds) for remote servers to refresh the trust bundle. Range: 60-3600 seconds.

* The `spec.federation.managedRoute` field enables automatic route creation by the Operator.

. Apply the configuration changes by running the following command:
+
[source,terminal]
----
$ oc apply -f spire-server.yaml
----

> . Check the status of the SPIRE Server by entering the following command. Wait for the `Ready` status to be returned.

Reviewer: ```suggestion
. Check the status of the SPIRE Server by entering the following command:
```

Author: Done

+
[source,terminal]
----
$ oc get spireserver cluster -w
----

> 

Reviewer: ```suggestion
+
Wait for the `Ready` status to be returned.
```

Author: Done

> . Verify that the federation route has been created:

Reviewer: ```suggestion
. Verify that the federation route has been created by running the following command:
```

Author: Done

+
[source,terminal]
----
$ oc get route -n zero-trust-workload-identity-manager | grep federation
----
+
.Example output
[source,terminal]
----
NAME                      HOST/PORT                                    PATH   SERVICES        PORT    TERMINATION
spire-server-federation   federation.apps.cluster1.example.com               spire-server     8443    passthrough
----

. Fetch the trust bundle from each remote cluster's federation endpoint:
+
[source,terminal]
----

> $  curl -k https://federation.apps.cluster2.example.com > cluster2-bundle.json

Reviewer: user replaceable value

Author: Done

----
+
[NOTE]
====

> For `https_spiffe` profile, you might need to use the `-k` flag if the certificate is not trusted by your system's CA bundle:

Reviewer: ```suggestion
For `https_spiffe` profile, you might need to use the `-k` flag if the certificate is not trusted by your system CA bundle:
```

Author: Done

> For `https_spiffe` profile, you might need to use the `-k` flag if the certificate is not trusted by your system's CA bundle:

Reviewer: might? is there some other way to run this in that case? or do we have to use the flag in that case?

Author: So I deleted "you might need to use" and it now says "profile, use the ..."

====
+
The response contains the trust bundle in JSON Web Key Set (JWKS) format:
+
.Example trust bundle
[source,json]
----
{
  "keys": [
    {
      "use": "x509-svid",
      "kty": "RSA",
      "n": "...",
      "e": "AQAB",
      "x5c": ["..."]
    }
  ],
  "spiffe_sequence": 1,
  "refresh_hint": 300
}
----

. Create `ClusterFederatedTrustDomain` resources for each remote trust domain.
+
.. On Cluster 1, create a resource to federate with Cluster 2:
+
[source,yaml]
----
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster2
spec:
  trustDomain: <CLUSTER2_APPS_DOMAIN>
  bundleEndpointURL: https://federation.<CLUSTER2_APPS_DOMAIN>
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://<CLUSTER2_APPS_DOMAIN>/spire/server
  className: zero-trust-workload-identity-manager-spire
  trustDomainBundle: |
    {
      "keys": [
        {
          "use": "x509-svid",
          "kty": "RSA",
          "n": "...",
          "e": "AQAB",
          "x5c": ["..."]
        }
      ],
      "spiffe_sequence": 1
    }
----
+

> where

Reviewer: ```suggestion
where:
```

Author: Done

`<CLUSTER2_APPS_DOMAIN>`:: Specifies the trust domain of the external cluster you are federating with.

`spec.bundleEndpointProfile.endpointSPIFFEID`:: Specifies the SPIFFE ID of the remote SPIRE server. Required for `https_spiffe` profile to validate the remote server's identity.

`spec.trustDomainBundle`:: Specifies the complete trust bundle JSON that you fetched in the previous step.

`spec.className`:: Specifies the name of a class to watch CRs for. Spire-controller-manager watches the resource only if `spec.className` is set to `zero-trust-workload-identity-manager-spire`.

. Apply the `ClusterFederatedTrustDomain` resource by running the following command:
+
[source,terminal]
----
$ oc apply -f clusterfederatedtrustdomain.yaml
----

> . Repeat steps 5-7 on each cluster for every remote cluster it should federate with. For bidirectional federation, each cluster needs a `ClusterFederatedTrustDomain` resource for every other cluster.

Reviewer: ```suggestion
. Repeat steps 5-7 on each cluster for every remote cluster that Cluster 1 should federate with. For bidirectional federation, each cluster needs a `ClusterFederatedTrustDomain` resource for every other cluster.
```
i think this is what you mean; by here i had forgotten if "it" was cluster 1 or a different one

Author: That is what I mean. Thanks.

. Update the `SpireServer` resource on each cluster to add the `federatesWith` configuration:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: SpireServer
metadata:
  name: cluster
spec:
  trustDomain: cluster1.example.com
  federation:
    bundleEndpoint:
      profile: https_spiffe
      refreshHint: 300
    federatesWith:
      - trustDomain: cluster2.example.com
        bundleEndpointUrl: https://federation.apps.cluster2.example.com
        bundleEndpointProfile: https_spiffe
        endpointSpiffeId: spiffe://cluster2.example.com/spire/server
      - trustDomain: cluster3.example.com
        bundleEndpointUrl: https://federation.apps.cluster3.example.com
        bundleEndpointProfile: https_spiffe
        endpointSpiffeId: spiffe://cluster3.example.com/spire/server
    managedRoute: "true"
----

> 

Reviewer: ```suggestion
+
```

Author: Done
